# Project description

## Project goals

***Testing changes:*** 
* The introduction of a new interface for an improved recommendation system for EU users.

***Test name:***
* interface_eu_test

***Audience:***
* The audience is new users from the EU who were encouraged to sign up in the online store. 
* Expected number of test participants: 6000

***Test duration:***
* From December 7 to December 21, 2020.

***Expected result:*** 
* Increase total conversion by at least 10% at each stage of the funnel `product_page → product_cart → purchase` in test group B.

## Data description

- `ab_project_marketing_events_us.csv` — the calendar of marketing events for 2020
- `final_ab_new_users_us.csv` — all users who signed up in the online store from December 7 to 21, 2020
- `final_ab_events_us.csv` — all events of the new users within the period from December 7, 2020 to January 1, 2021
- `final_ab_participants_us.csv` — table containing test participants

## Table of Contents

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Project-description" data-toc-modified-id="Project-description-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Project description</a></span><ul class="toc-item"><li><span><a href="#Project-goals" data-toc-modified-id="Project-goals-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Project goals</a></span></li><li><span><a href="#Data-description" data-toc-modified-id="Data-description-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Data description</a></span></li><li><span><a href="#Table-of-Contents" data-toc-modified-id="Table-of-Contents-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Table of Contents</a></span></li></ul></li><li><span><a href="#Project-implementation" data-toc-modified-id="Project-implementation-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Project implementation</a></span><ul class="toc-item"><li><span><a href="#Project-settings" data-toc-modified-id="Project-settings-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Project settings</a></span><ul class="toc-item"><li><span><a href="#Importing-libraries" data-toc-modified-id="Importing-libraries-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Importing libraries</a></span></li><li><span><a href="#Functions---helpers-:)" data-toc-modified-id="Functions---helpers-:)-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>Functions - helpers :)</a></span><ul class="toc-item"><li><span><a href="#eda_df()" data-toc-modified-id="eda_df()-2.1.2.1"><span class="toc-item-num">2.1.2.1&nbsp;&nbsp;</span>eda_df()</a></span></li><li><span><a href="#dtype_converter()" data-toc-modified-id="dtype_converter()-2.1.2.2"><span class="toc-item-num">2.1.2.2&nbsp;&nbsp;</span>dtype_converter()</a></span></li><li><span><a href="#time_diff()" data-toc-modified-id="time_diff()-2.1.2.3"><span class="toc-item-num">2.1.2.3&nbsp;&nbsp;</span>time_diff()</a></span></li><li><span><a href="#get_day()" data-toc-modified-id="get_day()-2.1.2.4"><span class="toc-item-num">2.1.2.4&nbsp;&nbsp;</span>get_day()</a></span></li><li><span><a href="#check_hypothesis()" data-toc-modified-id="check_hypothesis()-2.1.2.5"><span class="toc-item-num">2.1.2.5&nbsp;&nbsp;</span>check_hypothesis()</a></span></li></ul></li><li><span><a href="#Data-sources" data-toc-modified-id="Data-sources-2.1.3"><span class="toc-item-num">2.1.3&nbsp;&nbsp;</span>Data sources</a></span></li></ul></li><li><span><a href="#Data-Overview-and-Preprocessing" data-toc-modified-id="Data-Overview-and-Preprocessing-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Data Overview and Preprocessing</a></span><ul class="toc-item"><li><span><a href="#The-calendar_events-dataset" data-toc-modified-id="The-calendar_events-dataset-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>The <code>calendar_events</code> dataset</a></span></li><li><span><a href="#The-test_groups-dataset" data-toc-modified-id="The-test_groups-dataset-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>The <code>test_groups</code> dataset</a></span></li><li><span><a href="#The-new_users-dataset" data-toc-modified-id="The-new_users-dataset-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>The <code>new_users</code> dataset</a></span></li><li><span><a href="#The-events-dataset" data-toc-modified-id="The-events-dataset-2.2.4"><span class="toc-item-num">2.2.4&nbsp;&nbsp;</span>The <code>events</code> dataset</a></span></li></ul></li><li><span><a href="#Merging-datasets" data-toc-modified-id="Merging-datasets-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Merging datasets</a></span></li><li><span><a href="#Exploratory-Data-Analysis-(EDA)" data-toc-modified-id="Exploratory-Data-Analysis-(EDA)-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Exploratory Data Analysis (EDA)</a></span><ul class="toc-item"><li><span><a href="#Study-conversion-at-different-funnel-stages." data-toc-modified-id="Study-conversion-at-different-funnel-stages.-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Study conversion at different funnel stages.</a></span></li><li><span><a href="#Is-the-number-of-events-per-user-distributed-equally-in-the-samples?" data-toc-modified-id="Is-the-number-of-events-per-user-distributed-equally-in-the-samples?-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>Is the number of events per user distributed equally in the samples?</a></span></li><li><span><a href="#Are-there-users-who-enter-both-samples?" data-toc-modified-id="Are-there-users-who-enter-both-samples?-2.4.3"><span class="toc-item-num">2.4.3&nbsp;&nbsp;</span>Are there users who enter both samples?</a></span></li><li><span><a href="#How-is-the-number-of-events-distributed-by-days?" data-toc-modified-id="How-is-the-number-of-events-distributed-by-days?-2.4.4"><span class="toc-item-num">2.4.4&nbsp;&nbsp;</span>How is the number of events distributed by days?</a></span></li><li><span><a href="#Think-of-the-possible-details-in-the-data-that-you-have-to-take-into-account-before-starting-the-A/B-test?" data-toc-modified-id="Think-of-the-possible-details-in-the-data-that-you-have-to-take-into-account-before-starting-the-A/B-test?-2.4.5"><span class="toc-item-num">2.4.5&nbsp;&nbsp;</span>Think of the possible details in the data that you have to take into account before starting the A/B test?</a></span></li><li><span><a href="#Conclusion:" data-toc-modified-id="Conclusion:-2.4.6"><span class="toc-item-num">2.4.6&nbsp;&nbsp;</span>Conclusion:</a></span></li></ul></li><li><span><a href="#Evaluate-the-A/B-test-results" data-toc-modified-id="Evaluate-the-A/B-test-results-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Evaluate the A/B test results</a></span><ul class="toc-item"><li><span><a href="#What-can-you-tell-about-the-A/A-test-results?" data-toc-modified-id="What-can-you-tell-about-the-A/A-test-results?-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>What can you tell about the A/A test results?</a></span></li><li><span><a href="#Use-the-z-criterion-to-check-the-statistical-difference-between-the-proportions" data-toc-modified-id="Use-the-z-criterion-to-check-the-statistical-difference-between-the-proportions-2.5.2"><span class="toc-item-num">2.5.2&nbsp;&nbsp;</span>Use the z-criterion to check the statistical difference between the proportions</a></span></li></ul></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></li><li><span><a href="#General-conclusion" data-toc-modified-id="General-conclusion-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>General conclusion</a></span></li></ul></div>

# Project implementation

## Project settings

### Importing libraries

In [1]:
import os 
import sys

import pandas as pd
import numpy as np

import math
from datetime import datetime

from scipy import stats
from scipy.stats import mode

# visualization
import plotly.express as px
import plotly.io as pio
pio.templates.default = "none"

# to show warnings only once:
import warnings; warnings.filterwarnings(action='once')

# avoid warnings (if necessary)
import warnings
if not sys.warnoptions:
       warnings.simplefilter("ignore")

# to print all the outputs in the cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

### Functions - helpers :)

#### eda_df()

In [2]:
def eda_df(df):
    """
    Robust EDA function:
    1. Converts unordered categories to ordered to allow min/max operations.
    2. Separates numeric logic to prevent errors with mean/std/skew.
    """
    print('----------------- Summary statistics -----------------')

    # Pre-processing: Fix unordered categories
    for col in df.select_dtypes(include=['category']).columns:
        if not df[col].cat.ordered:
            df[col] = df[col].cat.as_ordered()

    # Separate numeric columns for specific math operations
    numeric_df = df.select_dtypes(include=[np.number])

    eda = pd.DataFrame({
        'dtype': df.dtypes,
        'count': df.count(),
        'nunique': df.nunique(),
        'null_sum': df.isnull().sum(),
        'null_%': round((df.isnull().mean() * 100), 2),
        'min': df.min(numeric_only=False), # Now safe for ordered categories
        'max': df.max(numeric_only=False), # Now safe for ordered categories
        'mean': numeric_df.mean(),
        'median': numeric_df.median(),
        'std': numeric_df.std(),
        'skew': numeric_df.skew()
    })

    return eda.fillna('-')

#### dtype_converter()

In [3]:
def dtype_converter(series, dtype):
    """
    The function converts data type of series to another.
    And returns series.
    
    """
    
    types_data = ['int', 'float', 'float64', 'bool', 'str', 'category']
    try:
        if any(elem in dtype for elem in types_data):
            series = series.astype(dtype)      
        elif dtype == 'date':
            series = pd.to_datetime(series, format='%Y-%m-%dT%H:%M:%S')
        elif dtype == 'numeric':
            pd.to_numeric(series, errors='ignore')
    except ValueError:
        pass

    return series

#### time_diff()

In [4]:
def time_diff(end_ts, start_ts, time_val):
    """
    The function returns a difference between two timestamp in time_val
    """
    
    try:
        if time_val == 'seconds':
            return (end_ts - start_ts).dt.seconds
        elif time_val == 'days':
            return (end_ts - start_ts).dt.days 
        elif time_val == 'week':
            return (end_ts - start_ts).dt.week 
        elif time_val == 'month':
            return (end_ts - start_ts).dt.month
        elif time_val == 'year':
            return (end_ts - start_ts).dt.year
        
    except ValueError:
        'Data type is not datetime64[ns]'

####  get_day()

In [5]:
def get_day(df, date_column): 
    """
    The function retrives day from the `date_column` and returns it in datetime64[ns] format.
    """
    
    return pd.to_datetime(df[date_column].dt.date)

#### check_hypothesis()

In [6]:
def check_hypothesis(sample1, sample2, event, alpha):

    # let's start with successes, using

    successes1 = users_pivot[users_pivot.event_name == event][sample1].iloc[0]
    successes2 = users_pivot[users_pivot.event_name == event][sample2].iloc[0]

    # for trials we can go back to original df or used a pre-aggregated data
    trials1 = df_ab_testing[df_ab_testing.group == sample1]['user_id'].nunique()
    trials2 = df_ab_testing[df_ab_testing.group == sample2]['user_id'].nunique()

    # proportion for success in the first group
    p1 = successes1 / trials1

    # proportion for success in the second group
    p2 = successes2 / trials2

    # proportion in a combined dataset
    p_combined = (successes1 + successes2) / (trials1 + trials2)

    difference = p1 - p2

    # calculating the statistic in standard deviations of the standard normal distribution
    z_value = difference / math.sqrt(p_combined * (1 - p_combined) * (1 / trials1 + 1 / trials2))

    # setting up the standard normal distribution (mean 0, standard deviation 1)
    distr = stats.norm(0, 1)

    p_value = (1 - distr.cdf(abs(z_value))) * 2

    print('p-value: ', p_value)

    if (p_value < alpha):
        print("Reject H0 for", event, 'and groups', sample1, sample2)
    else:
        print("Fail to Reject H0 for", event, 'and groups', sample1, sample2)

###  Data sources

In [7]:
# source_dir = '/datasets/'
source_dir = '../datasets/' # works for me:)

missing_values = ["n/a", "na", "--"]

# filenames
calendar_events = 'ab_project_marketing_events_us.csv'
test_groups = 'final_ab_participants_us.csv'
new_users = 'final_ab_new_users_us.csv'
events = 'final_ab_events_us.csv'

## Data Overview and Preprocessing

### The `calendar_events` dataset

*Downloading:*

In [8]:
df_calendar_events = pd.read_csv(os.path.join(source_dir, calendar_events),
                        parse_dates=['start_dt', 'finish_dt'],
                        na_values=missing_values)

# saving the original data set for comparing changes to be made
orig_df_calendar_events = df_calendar_events.copy()

*Overview and descriptive statistic:*

In [9]:
print('\n----------------- The number of rows and columns -----------------')
df_calendar_events.shape
print('\n----------------- A concise summary of the DataFrame -----------------\n')
df_calendar_events.info()

print('\n----------------- Dataset -----------------')
df_calendar_events = df_calendar_events.sort_values(by='start_dt')
df_calendar_events


----------------- The number of rows and columns -----------------


(14, 4)


----------------- A concise summary of the DataFrame -----------------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   name       14 non-null     object        
 1   regions    14 non-null     object        
 2   start_dt   14 non-null     datetime64[ns]
 3   finish_dt  14 non-null     datetime64[ns]
dtypes: datetime64[ns](2), object(2)
memory usage: 580.0+ bytes

----------------- Dataset -----------------


,name,regions,start_dt,finish_dt
6,Chinese New Year Promo,APAC,2020-01-25,2020-02-07
1,St. Valentine's Day Giveaway,"EU, CIS, APAC, N.America",2020-02-14,2020-02-16
8,International Women's Day Promo,"EU, CIS, APAC",2020-03-08,2020-03-10
2,St. Patric's Day Promo,"EU, N.America",2020-03-17,2020-03-19
3,Easter Promo,"EU, CIS, APAC, N.America",2020-04-12,2020-04-19
7,Labor day (May 1st) Ads Campaign,"EU, CIS, APAC",2020-05-01,2020-05-03
9,Victory Day CIS (May 9th) Event,CIS,2020-05-09,2020-05-11
11,Dragon Boat Festival Giveaway,APAC,2020-06-25,2020-07-01
4,4th of July Promo,N.America,2020-07-04,2020-07-11
13,Chinese Moon Festival,APAC,2020-10-01,2020-10-07


In [10]:
df_calendar_events['n_days'] = time_diff(df_calendar_events.finish_dt, df_calendar_events.start_dt, 'days')
df_calendar_events

,name,regions,start_dt,finish_dt,n_days
6,Chinese New Year Promo,APAC,2020-01-25,2020-02-07,13
1,St. Valentine's Day Giveaway,"EU, CIS, APAC, N.America",2020-02-14,2020-02-16,2
8,International Women's Day Promo,"EU, CIS, APAC",2020-03-08,2020-03-10,2
2,St. Patric's Day Promo,"EU, N.America",2020-03-17,2020-03-19,2
3,Easter Promo,"EU, CIS, APAC, N.America",2020-04-12,2020-04-19,7
7,Labor day (May 1st) Ads Campaign,"EU, CIS, APAC",2020-05-01,2020-05-03,2
9,Victory Day CIS (May 9th) Event,CIS,2020-05-09,2020-05-11,2
11,Dragon Boat Festival Giveaway,APAC,2020-06-25,2020-07-01,6
4,4th of July Promo,N.America,2020-07-04,2020-07-11,7
13,Chinese Moon Festival,APAC,2020-10-01,2020-10-07,6


> Preliminary overview:
* The `df_calendar_events` dataset contains 14 events over the year 2020.
* The longest event duration is 13 days and belongs to 'Chinese New Year Promo'.
* The shortest event is 1 day. This is 'Single's Day Gift Promo'.
* There are no marketing events in August and September.
* The 'CIS New Year Gift Lottery' event coincides with the 'Christmas&New Year Promo' event. They overlap from December 30 to January 3 inclusive.
* There is no null values.
* **The 14 days of A / B testing from December 7 to 21 did not coincide with any marketing event.**

### The `test_groups` dataset

*Downloading:*

In [11]:
df_test_groups = pd.read_csv(os.path.join(source_dir, test_groups),
                             na_values=missing_values,
                             dtype={
                                 'ab_test': 'category',
                                 'group': 'category'
                             })

# saving the original data set for comparing changes to be made
orig_df_test_groups = df_test_groups.copy()

*Overview and descriptive statistic:*

In [12]:
eda_df(df_test_groups)
print('\n----------------- Dataset head -----------------')
df_test_groups.head()
print('\n----------------- Dataset tail -----------------')
df_test_groups.tail()
print('\n----------------- Dataset sample -----------------')
df_test_groups.sample(5)
print('\n----------------- The number of rows, columns -----------------')
df_test_groups.shape

----------------- Summary statistics -----------------


,dtype,count,nunique,null_sum,null_%,min,max,mean,median,std,skew
ab_test,category,18268,2,0,0.0,interface_eu_test,recommender_system_test,-,-,-,-
group,category,18268,2,0,0.0,A,B,-,-,-,-
user_id,object,18268,16666,0,0.0,0002CE61FF2C4011,FFF58BC33966EB51,-,-,-,-



----------------- Dataset head -----------------


,user_id,group,ab_test
0,D1ABA3E2887B6A73,A,recommender_system_test
1,A7A3664BD6242119,A,recommender_system_test
2,DABC14FDDFADD29E,A,recommender_system_test
3,04988C5DF189632E,A,recommender_system_test
4,482F14783456D21B,B,recommender_system_test



----------------- Dataset tail -----------------


,user_id,group,ab_test
18263,1D302F8688B91781,B,interface_eu_test
18264,3DE51B726983B657,A,interface_eu_test
18265,F501F79D332BE86C,A,interface_eu_test
18266,63FBE257B05F2245,A,interface_eu_test
18267,79F9ABFB029CF724,B,interface_eu_test



----------------- Dataset sample -----------------


,user_id,group,ab_test
16896,8F3C5E6E7BCA75FE,B,interface_eu_test
13330,057833E9F648F7EF,A,interface_eu_test
14144,23444DB5ADA226A6,A,interface_eu_test
292,582E07EE1FDB2E08,B,recommender_system_test
6687,15E18335C5DB30C9,A,recommender_system_test



----------------- The number of rows, columns -----------------


(18268, 3)

*Count unique values in the columns:*

In [13]:
df_test_groups.apply(lambda x: len(x.unique()))

df_test_groups[['group', 'ab_test']].apply(lambda x: x.unique())

user_id    16666
group          2
ab_test        2
dtype: int64

,group,ab_test
0,A,recommender_system_test
1,B,interface_eu_test


In [14]:
df_test_groups.duplicated().sum()

np.int64(0)

> Preliminary overview:
* The `test_groups` dataset contains 18,268 records
* All users are divided into 2 test groups and 2 types of A / B tests.
* There are no duplicates and Nan values.
* However, there are noticeably fewer unique test participants (16,666) than entries. This means that we have users in the dataset who passed both tests, `recommender_system_test` and in `interface_eu_test`. 

*I think we need to delete users who participate in two tests because we don't know which version of the interface the participants in the second test saw.*

In [15]:
participants_of_2_tests = df_test_groups.groupby(['user_id'])['ab_test'].nunique().reset_index().query('ab_test == 2')
participants_of_2_tests.head()
participants_of_2_tests.shape[0]

df_test_groups.query('user_id == "0082295A41A867B5"')

,user_id,ab_test
2,001064FEAAB631A1,2
10,00341D8401F0F665,2
12,003B6786B4FF5B03,2
29,0082295A41A867B5,2
46,00E68F103C66C1F7,2


1602

,user_id,group,ab_test
4768,0082295A41A867B5,A,recommender_system_test
14161,0082295A41A867B5,B,interface_eu_test


*As we can see, the dataset contains 1602 records of users who participate in both tests. Let's remove them.*

In [16]:
df_test_groups.shape[0]

# Create a list of user ids to remove
lst_users_to_remove = participants_of_2_tests.user_id.to_list()

# Removing users who are in both A and B test groups
df_test_groups = df_test_groups[~df_test_groups.user_id.isin(lst_users_to_remove)]

# Verifying the result:
df_test_groups.query('user_id == "lst_users_to_remove[0]"')
df_test_groups.shape[0]

18268

,user_id,group,ab_test


15064

*The `test_groups` dataset has 2 types of A/B test:*
- recommender_system_test,
- interface_eu_test.

*For the purposes of our current project, we have to use data related to `interface_eu_test` test.* 

*So let's split the dataset to get the necessary data and study it.*

In [17]:
df_eu_test = df_test_groups.query('ab_test == "interface_eu_test"')
eda_df(df_eu_test)

print('\n----------------- Dataset head -----------------')
df_eu_test.head()
print('\n----------------- The number of rows, columns -----------------')
df_eu_test.shape

df_eu_test.groupby('group')['user_id'].nunique()

----------------- Summary statistics -----------------


,dtype,count,nunique,null_sum,null_%,min,max,mean,median,std,skew
ab_test,category,9965,1,0,0.0,interface_eu_test,interface_eu_test,-,-,-,-
group,category,9965,2,0,0.0,A,B,-,-,-,-
user_id,object,9965,9965,0,0.0,0002CE61FF2C4011,FFF58BC33966EB51,-,-,-,-



----------------- Dataset head -----------------


,user_id,group,ab_test
6701,D4E530F6595A05A3,A,interface_eu_test
6703,6BCB0F33D3BAB8C2,A,interface_eu_test
6707,2D2E7AA539CF348F,B,interface_eu_test
6708,4BA448BCE1343C6F,A,interface_eu_test
6709,76B6CDF2A8B1DBFC,B,interface_eu_test



----------------- The number of rows, columns -----------------


(9965, 3)

group
A    5012
B    4953
Name: user_id, dtype: int64

*Now let's check if the dataset has users who are in both test groups A and B.*

In [18]:
participants_of_2_group = df_eu_test.groupby(['user_id'])['group'].nunique().reset_index().query('group == 2')
participants_of_2_group.head()
participants_of_2_group.shape[0]

,user_id,group


0

In [19]:
tmp_by_group = df_eu_test.groupby('group')['user_id'].agg([
    ('n_users', 'count')
]).sort_values(by='n_users', ascending=False).reset_index()

tmp_by_group

fig = px.pie(
    tmp_by_group,
    values='n_users',
    names='group',
    #              color='region',
    title='The ratio of users of 2 test groups.')

fig.show()

,group,n_users
0,A,5012
1,B,4953


> Conclusion:
* After preprocessing, we have a dataset named `df_eu_test`, which contains the users and groups of data we need,` interface_eu_test` A / B test.
* The dataset `df_eu_test` has 9965 unique participants which are almost equally split:
      * 5012 users - in group A,
      * 4953 users - in group B
* In our dataset we do not have users that got into both test types or into both test groups.

### The `new_users` dataset

*Downloading:*

In [20]:
df_new_users = pd.read_csv(os.path.join(source_dir, new_users),
                           parse_dates=['first_date'],
                           na_values=missing_values,
                           dtype={
                               'region': 'category',
                               'device': 'category'
                           })

# saving the original data set for comparing changes to be made
orig_df_new_users = df_new_users.copy()

*Overview and descriptive statistic:*

In [21]:
eda_df(df_new_users)
print('\n----------------- Dataset head -----------------')
df_new_users.head()
print('\n----------------- Dataset tail -----------------')
df_new_users.tail()
print('\n----------------- Dataset sample -----------------')
df_new_users.sample(5)
print('\n----------------- The number of rows, columns -----------------')
df_new_users.shape

----------------- Summary statistics -----------------


,dtype,count,nunique,null_sum,null_%,min,max,mean,median,std,skew
device,category,61733,4,0,0.0,Android,iPhone,-,-,-,-
first_date,datetime64[ns],61733,17,0,0.0,2020-12-07 00:00:00,2020-12-23 00:00:00,-,-,-,-
region,category,61733,4,0,0.0,APAC,N.America,-,-,-,-
user_id,object,61733,61733,0,0.0,0001710F4DDB1D1B,FFFFE36C0F6E92DF,-,-,-,-



----------------- Dataset head -----------------


,user_id,first_date,region,device
0,D72A72121175D8BE,2020-12-07,EU,PC
1,F1C668619DFE6E65,2020-12-07,N.America,Android
2,2E1BF1D4C37EA01F,2020-12-07,EU,PC
3,50734A22C0C63768,2020-12-07,EU,iPhone
4,E1BDDCE0DAFA2679,2020-12-07,N.America,iPhone



----------------- Dataset tail -----------------


,user_id,first_date,region,device
61728,1DB53B933257165D,2020-12-20,EU,Android
61729,538643EB4527ED03,2020-12-20,EU,Mac
61730,7ADEE837D5D8CBBD,2020-12-20,EU,PC
61731,1C7D23927835213F,2020-12-20,EU,iPhone
61732,8F04273BB2860229,2020-12-20,EU,Android



----------------- Dataset sample -----------------


,user_id,first_date,region,device
46191,3EF3D1086F401E40,2020-12-12,EU,PC
4977,C967DB4C341410A5,2020-12-07,N.America,Mac
8133,BA848F02BEA5B7F9,2020-12-14,EU,Android
1026,52A9EB797A7F2F56,2020-12-07,EU,PC
53536,AB1A4B66131BDD8F,2020-12-13,EU,PC



----------------- The number of rows, columns -----------------


(61733, 4)

*Count unique values in the columns:*

In [22]:
df_new_users.apply(lambda x: len(x.unique()))

user_id       61733
first_date       17
region            4
device            4
dtype: int64

In [23]:
df_new_users[['region', 'device']].apply(lambda x: x.unique())

,region,device
0,EU,PC
1,N.America,Android
2,APAC,iPhone
3,CIS,Mac


*Checking duplicates:*

In [24]:
df_new_users.duplicated().sum()

np.int64(0)

> Preliminary overview:
* The `new_users` dataset contains 61733 unique entries.
* New users signed up in the online store from December 7 to 23, 2020.
* New users have come from 4 regions: EU, N.America, APAC, CIS.
* They utilize 4 types of devices: PC, Android, iPhone, Mac. 
* There are no duplicates and Nan values.

*To test the new interface of the online store, we have 2 mandatory testing conditions:*
* Testing lasted 14 days, from December 7 to December 21. So these days are included in the scope of the project;
* We have to research new EU users.

*Firstly, let's look at the days of December 22nd and 23rd as outside the scope of our task and remove them.*

In [25]:
df_new_users = df_new_users.query('first_date < "2020-12-22"')

# Verifying the result
df_new_users.first_date.max()

Timestamp('2020-12-21 00:00:00')

*Secondly, we have to extract data on new EU users from the dataset. Moreover, they must be our test participants.*

In [26]:
# List of new EU users who participate in our test
lst_eu_test_users = df_eu_test.user_id.to_list()

# Creating df of users who participate in our test
df_new_eu_users = df_new_users[(df_new_users['user_id'].isin(lst_eu_test_users)) & (df_new_users['region'] == "EU")]

*Let's explore the `df_new_eu_users` dataset:*

*Summary after clipping dataset:*

In [27]:
eda_df(df_new_eu_users)
print('\n----------------- Dataset head -----------------')
df_new_eu_users.head()
print('\n----------------- Dataset sample -----------------')
df_new_eu_users.sample(5)
print('\n----------------- The number of rows, columns -----------------')
df_new_eu_users.shape 

----------------- Summary statistics -----------------


,dtype,count,nunique,null_sum,null_%,min,max,mean,median,std,skew
device,category,8963,4,0,0.0,Android,iPhone,-,-,-,-
first_date,datetime64[ns],8963,15,0,0.0,2020-12-07 00:00:00,2020-12-21 00:00:00,-,-,-,-
region,category,8963,1,0,0.0,EU,EU,-,-,-,-
user_id,object,8963,8963,0,0.0,0002CE61FF2C4011,FFF58BC33966EB51,-,-,-,-



----------------- Dataset head -----------------


,user_id,first_date,region,device
2,2E1BF1D4C37EA01F,2020-12-07,EU,PC
3,50734A22C0C63768,2020-12-07,EU,iPhone
14,5BE017E9C8CC42F8,2020-12-07,EU,Android
18,96F27A054B191457,2020-12-07,EU,iPhone
21,E6AF85675078215D,2020-12-07,EU,Android



----------------- Dataset sample -----------------


,user_id,first_date,region,device
27716,21CE832F0DAFBC81,2020-12-09,EU,iPhone
9916,E7A9A15BBCB39D86,2020-12-14,EU,PC
37327,CCC24303EC961241,2020-12-17,EU,Android
3905,91E5060DF9C254D5,2020-12-07,EU,iPhone
9792,79F9ABFB029CF724,2020-12-14,EU,PC



----------------- The number of rows, columns -----------------


(8963, 4)

In [28]:
df_eu_test.shape

(9965, 3)

*The dataset of new EU users who signed up in the online store from December 7 to 21, 2020 contains fewer unique users than the dataset containing test participants:*
* dataset `df_new_eu_users`: 8963 users
* dataset ` df_eu_test`: 9965 users

*I can assume that some users took part in A / B testing without prior registration in the online store, for example, using the 'Buy Now' button, etc. We can check this version later.*

*So let's continue with our new EU users.*

In [29]:
tmp_users_by_date = df_new_eu_users.groupby('first_date')['user_id'].agg([
    ('n_users', 'count')]).reset_index()

tmp_users_by_date

fig = px.line(tmp_users_by_date,
              x="first_date",
              y="n_users",
              title='The daily number of newly signed users.',
              labels={
                  'n_users': 'Number of Users',
                  'first_date': 'Session Date'
              })

fig.show()

,first_date,n_users
0,2020-12-07,868
1,2020-12-08,515
2,2020-12-09,350
3,2020-12-10,464
4,2020-12-11,389
5,2020-12-12,694
6,2020-12-13,738
7,2020-12-14,886
8,2020-12-15,486
9,2020-12-16,324


* However, in fact, EU users make up 82% of our dataset.*

In [30]:
tmp_by_device = df_new_eu_users.groupby('device')['user_id'].agg([
    ('n_users', 'count')
]).sort_values(by='n_users', ascending=False).reset_index()

tmp_by_device

fig = px.pie(tmp_by_device,
             values='n_users',
             names='device',
#              color='region',
             title='The number of newly signed users by device.')

fig.show()

,device,n_users
0,Android,4022
1,PC,2281
2,iPhone,1730
3,Mac,930


> Conclusion:
* After preprocessing we have the `df_new_eu_users` dataset 8963 unique entries.
* All new users in the dataset are from the EU who signed up in the online store from December 7 to 21, 2020.
* New EU user registrations for 14 days of testing varied greatly:
    - The maximum number of new users came on December 21st - 1001 new users, 
    - The minimum - on December 16th -  324 new users,   
    - The number of new user registrations depends on the day of week.
* They utilize 4 types of devices: PC, Android, iPhone, Mac:
    - Almost half of the new users are Android users, 44.9%. 
    - The smallest number of users came using the Mac, 10.4%.
* There are no null values and duplicates in the dataset.
* **According to the assignment, the estimated number of test participants must be at least 6000. We have 8963 users, which meets the conditions of testing.**

### The `events` dataset

*Downloading:*

In [31]:
df_events = pd.read_csv(os.path.join(source_dir, events),
                        parse_dates=['event_dt'],
                        na_values=missing_values,
                        dtype={'event_name': 'category'})

# saving the original data set for comparing changes to be made
orig_df_events = df_events.copy()

*Overview and descriptive statistic:*

In [32]:
print('\n----------------- A concise summary of the DataFrame -----------------\n')
df_events.info()
print('\n----------------- Dataset head -----------------')
df_events.head()
print('\n----------------- Dataset tail -----------------')
df_events.tail()
print('\n----------------- Dataset sample -----------------')
df_events.sample(5)
print('\n----------------- The number of rows, columns -----------------')
df_events.shape


----------------- A concise summary of the DataFrame -----------------

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 423761 entries, 0 to 423760
Data columns (total 4 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   user_id     423761 non-null  object        
 1   event_dt    423761 non-null  datetime64[ns]
 2   event_name  423761 non-null  category      
 3   details     60314 non-null   float64       
dtypes: category(1), datetime64[ns](1), float64(1), object(1)
memory usage: 10.1+ MB

----------------- Dataset head -----------------


,user_id,event_dt,event_name,details
0,E1BDDCE0DAFA2679,2020-12-07 20:22:03,purchase,99.99
1,7B6452F081F49504,2020-12-07 09:22:53,purchase,9.99
2,9CD9F34546DF254C,2020-12-07 12:59:29,purchase,4.99
3,96F27A054B191457,2020-12-07 04:02:40,purchase,4.99
4,1FD7660FDF94CA1F,2020-12-07 10:15:09,purchase,4.99



----------------- Dataset tail -----------------


,user_id,event_dt,event_name,details
423756,245E85F65C358E08,2020-12-30 19:35:55,login,NaN
423757,9385A108F5A0A7A7,2020-12-30 10:54:15,login,NaN
423758,DB650B7559AC6EAC,2020-12-30 10:59:09,login,NaN
423759,F80C9BDDEA02E53C,2020-12-30 09:53:39,login,NaN
423760,7AEC61159B672CC5,2020-12-30 11:36:13,login,NaN



----------------- Dataset sample -----------------


,user_id,event_dt,event_name,details
222983,251941ECF8B3A4C6,2020-12-24 18:46:20,product_page,NaN
121033,1F74426A2674120A,2020-12-07 01:25:07,product_page,NaN
357966,A1C5ACC53B6CC6FF,2020-12-21 06:24:52,login,NaN
48775,135D554DB9C3B0E4,2020-12-23 03:59:01,purchase,4.99
231648,723943FFD9EBA3A1,2020-12-26 08:16:22,product_page,NaN



----------------- The number of rows, columns -----------------


(423761, 4)

*Count unique values in the columns:*

In [33]:
df_events.apply(lambda x: len(x.unique()))

user_id        58703
event_dt      257138
event_name         4
details            5
dtype: int64

In [34]:
for i in df_events[['event_name', 'details']]:
    print(i, ':', df_events[i].unique())

event_name : ['purchase', 'product_cart', 'product_page', 'login']
Categories (4, object): ['product_cart', 'product_page', 'purchase', 'login']
details : [ 99.99   9.99   4.99 499.99    nan]


*Let's explore Nan values in the `details` column:*

In [35]:
details_null = df_events.details.isna().sum()
details_null
(details_null / df_events.shape[0] * 100).round(2)

np.int64(363447)

np.float64(85.77)

*As we can see, our `event` dataset has 85.77% NaN values in the `details` column.
<br>I think this happened because not all of our users did reach the `purchase` event. We can check this version)*

In [36]:
df_events[(df_events.event_name == 'purchase') & df_events.details.isna()]

,user_id,event_dt,event_name,details


*There is no 'purchase' event with NaN value in the 'details' column. But previous steps of funnel have Nan's.*

*I consider that we can fill Nan's with 0.*

In [37]:
# filling Nan
df_events.details.fillna(0, inplace=True)

# Verifying the result
df_events.isna().sum()

user_id       0
event_dt      0
event_name    0
details       0
dtype: int64

*Checking duplicates:*

In [38]:
df_events.duplicated().sum()

np.int64(0)

In [39]:
df_events.event_dt.describe()

count                           423761
mean     2020-12-18 10:10:17.282395648
min                2020-12-07 00:00:33
25%                2020-12-14 03:05:18
50%                2020-12-18 17:40:52
75%                2020-12-22 13:51:17
max                2020-12-30 23:36:33
Name: event_dt, dtype: object

> Preliminary overview:
* The `events` dataset contains 423761 records from December 7 to December 30, 2020.
* The funnel consists 4 events: login, product_page, product_cart, and purchase. 
<br>However, in accordance with our task, we are interested in a funnel of these 3 steps: `product_page → product_card → purchase`. 
* There are no duplicates.
* Nan values of the variable `details` were filled with 0.

*To prepare the dataset for further exploration, we need to perform the same data transformation as for the new user dataset, namely:*
* First, get the data for our testing from December 7 to December 21;
* Second, we need to retrieve the events that were executed by our test participants.*

*Let's create the `df_test_events` for our goals and explore it:*

In [40]:
df_test_events = df_events[(df_events['user_id'].isin(lst_eu_test_users)) & (df_events['event_dt'] < "2020-12-22")]

eda_df(df_test_events)
print('\n----------------- Dataset sample -----------------')
df_test_events.sample(5)

----------------- Summary statistics -----------------


,dtype,count,nunique,null_sum,null_%,min,max,mean,median,std,skew
details,float64,51848,5,0,0.0,0.0,499.99,3.562788,0.0,29.740594,14.689667
event_dt,datetime64[ns],51848,33338,0,0.0,2020-12-07 00:02:48,2020-12-21 23:58:37,-,-,-,-
event_name,category,51848,4,0,0.0,product_cart,login,-,-,-,-
user_id,object,51848,8959,0,0.0,0002CE61FF2C4011,FFF58BC33966EB51,-,-,-,-



----------------- Dataset sample -----------------


,user_id,event_dt,event_name,details
303110,CFCAE736B9725AA6,2020-12-15 20:44:17,login,0.0
338062,ADB59CC4832BA191,2020-12-19 09:33:27,login,0.0
75624,FB2B7CA7C67C24F1,2020-12-14 14:57:33,product_cart,0.0
369433,51B099E406DE5300,2020-12-21 00:57:24,login,0.0
334139,7755F50B700473DC,2020-12-18 18:28:11,login,0.0


*Count unique values in the columns:*

In [41]:
df_test_events.apply(lambda x: len(x.unique()))

df_test_events[['event_name']].apply(lambda x: x.unique())

user_id        8959
event_dt      33338
event_name        4
details           5
dtype: int64

,event_name
0,purchase
1,product_cart
2,product_page
3,login


*Verifying that there are no duplicate values in the dataset:*

In [42]:
df_test_events.duplicated().sum()

np.int64(0)

*First look at the number of events per user:*

In [43]:
tmp_events = df_test_events.groupby('user_id')['event_dt'].agg([
    ('n_events', 'count')
]).sort_values(by='n_events', ascending=False).reset_index()

tmp_events.n_events.describe()

# most frequently number of events
tmp_events.n_events.mode()

count    8959.000000
mean        5.787253
std         3.727047
min         1.000000
25%         3.000000
50%         6.000000
75%         8.000000
max        24.000000
Name: n_events, dtype: float64

0    6
Name: n_events, dtype: int64

> Conclusion:
* The `df_test_events` dataset contains 51848 events.
* The dataset includes data that was collected from December 7 to December 21, 2020.
* All events presented in the dataset were performed by test takers.
* Each participant completed at least 1 event. Maximum events performed by user is 24.
* An average number of events per user is 6.
* Most often the number of events per user is 6, which coincides with the average number of events.

## Merging datasets

*After preprocessing our datasets, we can combine them for deeper exploratory analysis.
<br>But first, a small reminder of which ones we will use and what they look like :)* 

In [44]:
print('\n-----------------  Table containing test participants   -----------------')
df_eu_test.head()
df_eu_test.shape

print('\n----------------- New EU users who signed up in the online store from December 7 to 21, 2020 -----------------')
df_new_eu_users.head()
df_new_eu_users.shape

print('\n----------- All the events of the new users (test participants) within the period from December 7 to 21, 2020 ------------')
df_test_events.head()
df_test_events.shape


-----------------  Table containing test participants   -----------------


,user_id,group,ab_test
6701,D4E530F6595A05A3,A,interface_eu_test
6703,6BCB0F33D3BAB8C2,A,interface_eu_test
6707,2D2E7AA539CF348F,B,interface_eu_test
6708,4BA448BCE1343C6F,A,interface_eu_test
6709,76B6CDF2A8B1DBFC,B,interface_eu_test


(9965, 3)


----------------- New EU users who signed up in the online store from December 7 to 21, 2020 -----------------


,user_id,first_date,region,device
2,2E1BF1D4C37EA01F,2020-12-07,EU,PC
3,50734A22C0C63768,2020-12-07,EU,iPhone
14,5BE017E9C8CC42F8,2020-12-07,EU,Android
18,96F27A054B191457,2020-12-07,EU,iPhone
21,E6AF85675078215D,2020-12-07,EU,Android


(8963, 4)


----------- All the events of the new users (test participants) within the period from December 7 to 21, 2020 ------------


,user_id,event_dt,event_name,details
3,96F27A054B191457,2020-12-07 04:02:40,purchase,4.99
9,A92195E3CFB83DBD,2020-12-07 00:32:07,purchase,4.99
11,354D653172FF2A2D,2020-12-07 15:45:11,purchase,4.99
12,7FCD34F47C13A9AC,2020-12-07 22:06:13,purchase,9.99
13,0313C457F07C339E,2020-12-07 13:10:48,purchase,9.99


(51848, 4)

*Combining the datasets:*

In [45]:
df_ab_testing = df_test_events.merge(df_eu_test[['user_id', 'group']],
                                     on='user_id', how='left').merge(df_new_eu_users[[
                                         'user_id', 'first_date', 'device'
                                     ]], on='user_id', how='left')

df_ab_testing.head()

,user_id,event_dt,event_name,details,group,first_date,device
0,96F27A054B191457,2020-12-07 04:02:40,purchase,4.99,B,2020-12-07,iPhone
1,A92195E3CFB83DBD,2020-12-07 00:32:07,purchase,4.99,A,2020-12-07,Android
2,354D653172FF2A2D,2020-12-07 15:45:11,purchase,4.99,A,2020-12-07,Mac
3,7FCD34F47C13A9AC,2020-12-07 22:06:13,purchase,9.99,B,2020-12-07,PC
4,0313C457F07C339E,2020-12-07 13:10:48,purchase,9.99,A,2020-12-07,Mac


In [46]:
# Checking some features after merging
df_ab_testing.isna().sum()
df_ab_testing.duplicated().sum()

user_id       0
event_dt      0
event_name    0
details       0
group         0
first_date    0
device        0
dtype: int64

np.int64(0)

## Exploratory Data Analysis (EDA)

### Study conversion at different funnel stages.

*The main goal of our test is to make sure that the total conversion [1] on each stage in Test Group B is better than the total conversion at the same stage in Control Group A.*

[1] 'Total conversion is the ratio of users at a particular stage to the number of users at the first stage.'

*Let's plot the funnel and compare the groups.*

1) *Calculate the number of users for each stage in group A:*

In [47]:
df_funnel_A = df_ab_testing[df_ab_testing['group'] == "A"].groupby(
    ['event_name'])['user_id'].agg([
        ('n_users', 'count')
    ]).sort_values(by=['n_users'], ascending=False).reset_index()

df_funnel_A

,event_name,n_users
0,login,11247
1,product_page,7515
2,purchase,4031
3,product_cart,3720


2) *Calculating total conversion for each funnel stage in group A:*

In [48]:
first_stage_A = df_funnel_A.iloc[0, 1]
first_stage_A

np.int64(11247)

In [49]:
df_funnel_A = df_funnel_A.assign(
    total_conversion=lambda x: round(x.n_users / first_stage_A * 100, 2),
    test_group='A')

df_funnel_A

,event_name,n_users,total_conversion,test_group
0,login,11247,100.00,A
1,product_page,7515,66.82,A
2,purchase,4031,35.84,A
3,product_cart,3720,33.08,A


*The same steps for the group B:*

In [50]:
df_funnel_B = df_ab_testing[df_ab_testing['group'] == "B"].groupby(
    ['event_name'])['user_id'].agg([
        ('n_users', 'count')
    ]).sort_values(by=['n_users'], ascending=False).reset_index()

# df_funnel_B

first_stage_B = df_funnel_B.iloc[0, 1]
first_stage_B

df_funnel_B = df_funnel_B.assign(
    total_conversion=lambda x: round(x.n_users / first_stage_B * 100, 2),
    test_group='B')

df_funnel_B

np.int64(10885)

,event_name,n_users,total_conversion,test_group
0,login,10885,100.00,B
1,product_page,7102,65.25,B
2,product_cart,3722,34.19,B
3,purchase,3626,33.31,B


*Concatenating test groups and plotting funnel chart:*

In [51]:
df_funnel_shift = pd.concat([df_funnel_A, df_funnel_B], axis=0)
df_funnel_shift

fig = px.funnel(df_funnel_shift,
                x='total_conversion',
                y='event_name',
                color='test_group',
                labels={
                    'event_name': ' ',
                    'test_group': 'Test groups:'
                })
fig.show()

,event_name,n_users,total_conversion,test_group
0,login,11247,100.00,A
1,product_page,7515,66.82,A
2,purchase,4031,35.84,A
3,product_cart,3720,33.08,A
0,login,10885,100.00,B
1,product_page,7102,65.25,B
2,product_cart,3722,34.19,B
3,purchase,3626,33.31,B


>* According to the test, it was expected that 'At each stage of the funnel `product_page → product_cart → purchase`, there will be at least a 10% increase."
* However, the funnel chart shows that the conversion in the control group A is even slightly higher than in the test group B in the first two stages.
* Only the `purchase` stage showed 1.11% better in test group B than in group A.
* Moreover, at `purchase` stage in group B, the total conversion is higher than at the previous stage. I assume this was due to the 'Buy Now' button, which allowed users to buy by skipping some steps.

### Is the number of events per user distributed equally in the samples?

In [52]:
tm_events_per_user = df_ab_testing.groupby(['group']).agg({
    'user_id': 'nunique',
    'event_dt': 'count'
}).rename(columns={
    'user_id': 'n_users',
    'event_dt': 'n_events'
}).reset_index()

tm_events_per_user['n_events_per_user'] = round(tm_events_per_user.n_events / tm_events_per_user.n_users, 2)
tm_events_per_user

fig = px.pie(
    tm_events_per_user,
    values='n_events',
    names='group',
    #              color='region',
    title='Distribution of the number of events per user.')

fig.show()

,group,n_users,n_events,n_events_per_user
0,A,4521,26513,5.86
1,B,4438,25335,5.71


>*As we can see, the share of users and events is close to 50/50.
<br>Thus, we can claim that the number of events per user distributed equally in the samples*

### Are there users who enter both samples?

In [53]:
len(
    df_ab_testing.groupby(['user_id']).agg({
        'group': 'nunique'
    }).reset_index().query('group == 2'))

0

>*There is no users who enter both samples.*

### How is the number of events distributed by days?

*To answer this question, let's create an additional column `event-date`:*

In [54]:
df_ab_testing['event_date'] = get_day(df_ab_testing, 'event_dt')
df_ab_testing.head(2)

,user_id,event_dt,event_name,details,group,first_date,device,event_date
0,96F27A054B191457,2020-12-07 04:02:40,purchase,4.99,B,2020-12-07,iPhone,2020-12-07
1,A92195E3CFB83DBD,2020-12-07 00:32:07,purchase,4.99,A,2020-12-07,Android,2020-12-07


>*Number of total daily events:*

In [55]:
tm_events_per_day = df_ab_testing.groupby('event_date')['event_name'].agg([
    ('n_events', 'count')
]).reset_index()
tm_events_per_day.head()

fig = px.line(tm_events_per_day,
              x="event_date",
              y="n_events",
              title='Number of total daily events.',
              labels={
                  'n_events': 'Number of events',
                  'event_date': 'Event Date'
              })

fig.show()

,event_date,n_events
0,2020-12-07,1883
1,2020-12-08,2168
2,2020-12-09,2156
3,2020-12-10,2399
4,2020-12-11,2465


>*Number of total daily events per test groups:*

In [56]:
tm_events_per_group = df_ab_testing.groupby(['event_date', 'group'])['event_dt'].agg([
    ('n_events', 'count')
]).reset_index()
tm_events_per_group.head()

fig = px.line(tm_events_per_group,
              x="event_date",
              y="n_events",
              color='group',
              title='Number of total daily events per group.',
              labels={
                  'n_events': 'Number of events',
                  'event_date': 'Event Date',
                  'group': 'Test groups:'
              })

fig.show()

,event_date,group,n_events
0,2020-12-07,A,955
1,2020-12-07,B,928
2,2020-12-08,A,1070
3,2020-12-08,B,1098
4,2020-12-09,A,1084


>*Number of total daily events by event name:*

In [57]:
tm_events_by_name = df_ab_testing.groupby(['event_date', 'event_name'])['event_dt'].agg([
    ('n_events', 'count')
]).reset_index()
tm_events_by_name.head()

fig = px.line(tm_events_by_name,
              x="event_date",
              y="n_events",
              color='event_name',
              title='Number of total daily events by event name.',
              labels={
                  'n_events': 'Number of events',
                  'event_date': 'Event Date',
                  'event_name': 'Events:'
              })

fig.show()

,event_date,event_name,n_events
0,2020-12-07,product_cart,250
1,2020-12-07,product_page,535
2,2020-12-07,purchase,281
3,2020-12-07,login,817
4,2020-12-08,product_cart,295


In [58]:
tm_events_by_device = df_ab_testing.groupby(['event_date', 'device'])['event_dt'].agg([
    ('n_events', 'count')
]).reset_index()
tm_events_by_device.head()

fig = px.line(tm_events_by_device,
              x="event_date",
              y="n_events",
              color='device',
              title='Number of total daily events per device.',
              labels={
                  'n_events': 'Number of events',
                  'event_date': 'Event Date',
                  'device': 'Devices:'
              })

fig.show()

,event_date,device,n_events
0,2020-12-07,Android,808
1,2020-12-07,Mac,237
2,2020-12-07,PC,480
3,2020-12-07,iPhone,358
4,2020-12-08,Android,979


>* The total number of events during the test increased significantly, 3 times.
* This feature is typical for both Groups A and B, for each event name and for number of events per device.
* Most of the events were performed by Android users, with the smallest user group being Mac fans.
* It is clear from the graphs that we can say that there were spikes of events on a certain weekday. In our case, it was on Thursdays.
* To funnel stages, `product_cart` and `purchase` have almost the same number of events during the test weeks. **I can assume that the abandoned cart percentage is very low, or the 'Buy Now' button works better than the whole funnel.**

### Think of the possible details in the data that you have to take into account before starting the A/B test?

>* The datasets had users who were assigned to multiple groups and who performed both tests at the same time. We have processed our data to avoid exposure to these factors.
* We also checked if the users are equally split.
* But there is an external factor that influences the test result. I consider that we have to take into account that the A / B test was carried out on the eve of Christmas. 
<br>And even if the specific testing time did not last at the same time as the marketing activities, we cannot trust the results 100% due to the well-known increase in consumer activity before the Christmas holidays.

### Conclusion:

>* The test group B did not achieve expected result of increasing total conversion at least 10% on each stage of the funnel `product_page → product_cart → purchase`.
* The test group B showed better result on the `purchase` stage.
* There is no users who enter both samples.
* The number of events per user distributed equally in the samples.

## Evaluate the A/B test results

### What can you tell about the A/A test results?

In [59]:
df_ab_testing.event_name.dtypes
df_ab_testing.group.dtypes

# changing dtype, to avoid issue with CategoricalIndex columns in pandas
# https://github.com/pandas-dev/pandas/issues/19136

df_ab_testing.event_name = dtype_converter(df_ab_testing.event_name, 'str')
df_ab_testing.group = dtype_converter(df_ab_testing.group, 'str')

CategoricalDtype(categories=['product_cart', 'product_page', 'purchase', 'login'], ordered=True, categories_dtype=object)

CategoricalDtype(categories=['A', 'B'], ordered=True, categories_dtype=object)

In [60]:
users_pivot = df_ab_testing.pivot_table(index='event_name',
                         values='user_id',
                         columns='group',
                         aggfunc=lambda x: x.nunique()).reset_index()

users_pivot= users_pivot.assign(diff_n_users=lambda x: x.A - x.B)
users_pivot

group,event_name,A,B,diff_n_users
0,login,4520,4438,82
1,product_cart,1479,1516,-37
2,product_page,3030,2913,117
3,purchase,1612,1479,133


In [61]:
users_crosstab = pd.crosstab(index=df_ab_testing.event_name,
                             columns=df_ab_testing.group,
                             values=df_ab_testing.user_id,
                             aggfunc=lambda x: x.nunique(),
                             normalize=True, # Normalize by dividing all values by the sum of values.
                             margins=True,
                             margins_name='Total').reset_index()

users_crosstab = users_crosstab.assign(diff_share_stage=lambda x: round((x.A - x.B )* 100, 2))
users_crosstab

group,event_name,A,B,Total,diff_share_stage
0,login,0.215371,0.211464,0.426836,0.39
1,product_cart,0.070472,0.072235,0.142707,-0.18
2,product_page,0.144375,0.138800,0.283175,0.56
3,purchase,0.076809,0.070472,0.147282,0.63
4,Total,0.504632,0.495368,1.000000,0.93


> We can argue that all the criteria we have for a successful A/A test:
* As can be seen from the aggregated tables, there is no significant difference between the samples. On Each stage the difference is less then 1%.
* For all groups, data on the same event is recorded.
* Users remain within their groups until the end of the test. 
* Users who may see different versions of the page during the test were removed from the groups.

### Use the z-criterion to check the statistical difference between the proportions

***H0: there is not a statistically significant difference in conversion between the samples A and B***

***H1: there is a statistically significant difference in conversion between the samples A and B***

In [62]:
for i in users_pivot.event_name.unique():
    check_hypothesis("A", "B", i, alpha=0.05)

p-value:  0.321766940707354
Fail to Reject H0 for login and groups A B
p-value:  0.14702821950355438
Fail to Reject H0 for product_cart and groups A B
p-value:  0.16608962447812559
Fail to Reject H0 for product_page and groups A B
p-value:  0.02036540675054077
Reject H0 for purchase and groups A B


## Conclusion
>* As a result of checking the hypothesis, we cannot reject H0 for all the events. 
* Thus, we can accept the equality of the proportions of the population of `A` and `B` groups. It means that the groups were split properly.

# General conclusion

> ***Datasets:***

* The logs were collected in several tables for the period from December 7 to December 30, 2020.
* The two types of tests were run simultaneously and overlapped using the same test participants.

> ***New interface for EU users test:***

* The test was carried out within 14 days, from December 7 to December 21, 2020.
* After preprocessing, we got a dataset that contains 51848 different events performed by 8959 unique users.
* Users are split equally into two groups:
        - `A` control group - 4521 users
        - `B` test group - 4438 users
* Control group `A` gets the old site version and one test group `B` gets the new ones:        

> ***The funnel:***

* Consists of 4 events: `login -> product_page -> product_cart -> purchase`
* The number of events per user distributed equally in the samples.

> ***Performing tests:***

* The A/A test shows that the control groups were split properly.
* A / B tests show that we cannot reject null hypotheses. In other words, there are no statistically significant differences in conversions between two groups across all 4 events.

> ***External factors:***

* The A / B test was carried out on the eve of Christmas during the period of increased buying activity before the Christmas holidays. Thus, we cannot trust the results 100%.


> ***Test result:***

* The test group B did not achieve the expected result of increasing total conversion at least 10% on each stage of the funnel `product_page → product_cart → purchase`.
* Moreover, total conversion in group B is almost the same or less than total conversion in group A on stages product_page -> product_cart.
* Only at the last stage,`purchase`, test group B showed a 1.11% better result than in group A and than at the `product_cart` stage in their group.
<br>It means that some EU users made purchasing in one-click using the 'Buy Now' button.


> ***Recommendation:***

* It can be assumed that **EU users prefer to shop in the shortest and fastest way**, skipping stages in the usual sales funnel order. This hypothesis should be tested.

***Thank you in advance for the reviewing*** 😊